src/app/ReadClient.cpp

```diff
@@ -596,18 +592,19 @@ void ReadClient::CancelLivenessCheckTimer()
 void ReadClient::OnLivenessTimeoutCallback(System::Layer * apSystemLayer, void * apAppState)
 {
     ReadClient * const client = reinterpret_cast<ReadClient *>(apAppState);
-    if (client->IsFree())
-    {
-        ChipLogError(DataManagement,
-                     "ReadClient::OnLivenessTimeoutCallback invoked on a free client! This is a bug in CHIP stack!");
-        return;
-    }
+
+    //
+    // Might as well try to see if this instance exists in the tracked list in the IM.
+    // This might blow-up if there is a bug (in which case), but that's ok since it's at least blowing
+    // up here and folks debugging would know why
+    //
+    VerifyOrDie(client->mpImEngine->InActiveReadClientList(client));
```

Comment:
This seems like it would blow up if we landed here after IM engine shutdown....  Maybe it's OK in the sense that this really should not happen.

Reply:
There are a couple of ways it could blow up - the readClient object may itself have been reclaimed, at which point, the mpEngine pointer would be invalid and it would likely segfault OR, as you said, the imEngine has since been destructed, at which point, it'll likely just fail.

In all cases, having it consolidated here at least provides clues for the person debugging.

Will update the comments to that effect.